## SF crime data analysis and modeling

### In this notebook, I used Spark SQL for big data analysis on SF crime data. (https://data.sfgov.org/Public-Safety/Police-Department-Incident-Reports-Historical-2003/tmnf-yvry).

In [3]:
# import package
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings

import os
os.environ["PYSPARK_PYTHON"] = "python3"

In [4]:
# import urllib.request
# urllib.request.urlretrieve("https://data.sfgov.org/api/views/tmnf-yvry/rows.csv?accessType=DOWNLOAD", "/tmp/myxxxx.csv")
# dbutils.fs.mv("file:/tmp/myxxxx.csv", "dbfs:/laioffer/spark_hw1/data/sf_03_18.csv")
# display(dbutils.fs.ls("dbfs:/laioffer/spark_hw1/data/"))

# https://data.sfgov.org/api/views/tmnf-yvry/rows.csv?accessType=DOWNLOAD


path name size dbfs:/laioffer/spark_hw1/data/sf_03_18.csv sf_03_18.csv 550859737

In [5]:
data_path = 'dbfs:/laioffer/spark_hw1/data/sf_03_18.csv'
# use this file name later

In [6]:
# read data from the data storage
crime_data_lines = sc.textFile(data_path)

# prepare data 
df_crimes = crime_data_lines.map(lambda line: [x.strip('"') for x in next(reader([line]))])

# get header
header = df_crimes.first()
header

# remove the first line of data
crimes = df_crimes.filter(lambda x: x != header)

# get the first line of data
# display(crimes.take(3))

# get the total number of data 
crimes.count()

Out[5]: 2215024

### Solove  big data issues via Spark
-  I've learned the following approaches, and I'll mainly use SQL and Dataframe for my analysis
-  approach 1: use RDD
-  approach 2: use Dataframe, register the RDD to a dataframe 
-  approach 3: use SQL

In [8]:
# Get dataframe and sql
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

df_opt1 = spark.read.format("csv").option("header", "true").load(data_path)
display(df_opt1)
df_opt1.createOrReplaceTempView("sf_crime")


IncidntNum Category Descript DayOfWeek Date Time PdDistrict Resolution Address X Y Location PdId SF Find Neighborhoods Current Police Districts Current Supervisor Districts Analysis Neighborhoods DELETE - Fire Prevention Districts DELETE - Police Districts DELETE - Supervisor Districts DELETE - Zip Codes DELETE - Neighborhoods DELETE - 2017 Fix It Zones Civic Center Harm Reduction Project Boundary Fix It Zones as of 2017-11-06 DELETE - HSOC Zones Fix It Zones as of 2018-02-07 CBD, BID and GBD Boundaries as of 2017 Areas of Vulnerability, 2016 Central Market/Tenderloin Boundary Central Market/Tenderloin Boundary Polygon - Updated HSOC Zones as of 2018-06-05 OWED Public Spaces 146196161 NON-CRIMINAL LOST PROPERTY Tuesday 09/23/2014 01:00 SOUTHERN NONE 800 Block of BRYANT ST -122.403404791479 37.775420706711 POINT (-122.403404791479 37.775420706711) 14619616171000 32 1 10 34 14 2 9 28853 34 null null null null null null 2 null null null null 150045675 ASSAULT BATTERY Thursday 01/15/2015 17:00 TARAVAL NONE 1800 Block of VICENTE ST -122.48560378101 37.7388214326705 POINT (-122.48560378101 37.7388214326705) 15004567504134 40 10 7 35 1 8 3 29491 35 null null null null null null 1 null null null null 140632022 SUSPICIOUS OCC INVESTIGATIVE DETENTION Wednesday 07/30/2014 09:32 BAYVIEW NONE 100 Block of GILLETTE AV -122.396535107224 37.7106603302503 POINT (-122.396535107224 37.7106603302503) 14063202264085 89 2 9 1 10 3 8 309 1 null null null null null null 1 null null null null 150383259 ASSAULT BATTERY Saturday 05/02/2015 23:10 BAYVIEW ARREST, BOOKED 2400 Block of PHELPS ST -122.400130573297 37.7300925390327 POINT (-122.400130573297 37.7300925390327) 15038325904134 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 040753980 OTHER OFFENSES RECKLESS DRIVING Friday 07/02/2004 13:43 BAYVIEW NONE I-280 / CESAR CHAVEZ ST -120.5 90 POINT (-120.5 90) 4075398065020 null null null null null null null null null null null null null null null null null null null null 040855122 SUICIDE SUICIDE BY JUMPING Tuesday 07/27/2004 15:19 SOUTHERN NONE 500 Block of I-80 -122.386667033903 37.7898821569191 POINT (-122.386667033903 37.7898821569191) 4085512260170 null null null null null null null 28856 null null null null null null null null null null null null 066085191 NON-CRIMINAL LOST PROPERTY Sunday 11/19/2006 17:45 BAYVIEW NONE 0 Block of GIANTS DR -122.38750147945 37.716878646429 POINT (-122.38750147945 37.716878646429) 6608519171000 88 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 050908404 VEHICLE THEFT STOLEN AUTOMOBILE Saturday 08/13/2005 17:00 TENDERLOIN NONE JENNINGS CT / INGALLS ST -120.5 90 POINT (-120.5 90) 5090840407021 null null null null null null null null null null null null null null null null null null null null 090768064 ARSON ARSON OF A VEHICLE Tuesday 07/28/2009 23:26 BAYVIEW NONE SELBY ST / OAKDALE AV -122.399686082806 37.739901780585 POINT (-122.399686082806 37.739901780585) 9076806426031 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 111027676 ASSAULT BATTERY Saturday 12/24/2011 07:00 SOUTHERN NONE 0 Block of DORE ST -122.412933062384 37.7739274524819 POINT (-122.412933062384 37.7739274524819) 11102767604134 32 1 10 34 8 2 9 28853 34 null 1 null 1 null null 2 null null 1 null 060508925 ASSAULT BATTERY Saturday 05/13/2006 11:26 NORTHERN NONE POLK ST / CEDAR ST -122.419834938618 37.7863580362743 POINT (-122.419834938618 37.7863580362743) 6050892504134 50 4 10 36 13 9 9 28858 36 null null null null null 10 2 null null null null 110318854 FRAUD FRAUDULENT CREDIT APPLICATION Thursday 04/14/2011 17:30 BAYVIEW NONE 100 Block of TOPEKA AV -122.398951090192 37.7338473162583 POINT (-122.398951090192 37.7338473162583) 11031885409340 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 131007921 SUSPICIOUS OCC SUSPICIOUS OCCURRENCE Thursday 11/28/2013 12:00 INGLESIDE NONE 1600 Block of DIAMOND ST -122.435709983124 37.7443688410754 POINT (-122.4357099831

In [9]:
# Use a Spark program that counts the number of crimes for different category.
# q1_result = df_opt1.groupBy('category').count().orderBy('count', ascending=False)
# display(q1_result)

# Spark SQL based
crimeCategory = spark.sql("SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC")
display(crimeCategory)

category Count LARCENY/THEFT 480448 OTHER OFFENSES 309358 NON-CRIMINAL 238323 ASSAULT 194694 VEHICLE THEFT 126602 DRUG/NARCOTIC 119628 VANDALISM 116059 WARRANTS 101379 BURGLARY 91543 SUSPICIOUS OCC 80444 MISSING PERSON 64961 ROBBERY 55867 FRAUD 41542 SECONDARY CODES 25831 FORGERY/COUNTERFEITING 23050 WEAPON LAWS 22234 TRESPASS 19449 PROSTITUTION 16701 STOLEN PROPERTY 11891 SEX OFFENSES, FORCIBLE 11742 DISORDERLY CONDUCT 10040 DRUNKENNESS 9826 RECOVERED VEHICLE 8716 DRIVING UNDER THE INFLUENCE 5672 KIDNAPPING 5346 RUNAWAY 4440 LIQUOR LAWS 4083 ARSON 3931 EMBEZZLEMENT 2988 LOITERING 2430 SUICIDE 1292 FAMILY OFFENSES 1183 BAD CHECKS 925 BRIBERY 813 EXTORTION 741 SEX OFFENSES, NON FORCIBLE 431 GAMBLING 348 PORNOGRAPHY/OBSCENE MAT 59 TREA 14

In [10]:
# Top-5 most frequent crimes in SF
crimes_pd_df = crimeCategory.toPandas()
display(crimes_pd_df)

#We can see the top-5 frequency crimes are:
#1.Larcency/theft
#2.Other offenses
#3.Non-criminal
#4.Assault
#5.vehicle theft

category Count LARCENY/THEFT 480448 OTHER OFFENSES 309358 NON-CRIMINAL 238323 ASSAULT 194694 VEHICLE THEFT 126602 DRUG/NARCOTIC 119628 VANDALISM 116059 WARRANTS 101379 BURGLARY 91543 SUSPICIOUS OCC 80444 MISSING PERSON 64961 ROBBERY 55867 FRAUD 41542 SECONDARY CODES 25831 FORGERY/COUNTERFEITING 23050 WEAPON LAWS 22234 TRESPASS 19449 PROSTITUTION 16701 STOLEN PROPERTY 11891 SEX OFFENSES, FORCIBLE 11742 DISORDERLY CONDUCT 10040 DRUNKENNESS 9826 RECOVERED VEHICLE 8716 DRIVING UNDER THE INFLUENCE 5672 KIDNAPPING 5346 RUNAWAY 4440 LIQUOR LAWS 4083 ARSON 3931 EMBEZZLEMENT 2988 LOITERING 2430 SUICIDE 1292 FAMILY OFFENSES 1183 BAD CHECKS 925 BRIBERY 813 EXTORTION 741 SEX OFFENSES, NON FORCIBLE 431 GAMBLING 348 PORNOGRAPHY/OBSCENE MAT 59 TREA 14

In [11]:
# Use a Spark program that count the number of crimes for different district, and visualize the results
# q2_result = df_opt1.groupBy('PdDistrict').count().orderBy('count', ascending=False)
# display(q2_result)

# Spark SQL based
crimeDistrict = spark.sql("SELECT PdDistrict , COUNT(*) AS Count FROM sf_crime GROUP BY PdDistrict ORDER BY Count DESC")
display(crimeDistrict)


PdDistrict Count SOUTHERN 399785 MISSION 300076 NORTHERN 272713 CENTRAL 226255 BAYVIEW 221000 INGLESIDE 194180 TENDERLOIN 191746 TARAVAL 166971 PARK 125479 RICHMOND 116818 null 1

In [12]:
# Number of crimes in different districts
crimeDistrict = crimeDistrict.toPandas()
display(crimeDistrict)

PdDistrict Count SOUTHERN 399785 MISSION 300076 NORTHERN 272713 CENTRAL 226255 BAYVIEW 221000 INGLESIDE 194180 TENDERLOIN 191746 TARAVAL 166971 PARK 125479 RICHMOND 116818 null 1

In [13]:
# Count the number of crimes each 'Sunday' at "SF downtown" 
# SF downtown range: X (-122.4213,-122.4313), Y(37.7540,37.7740)
crimeSunSFDT = spark.sql("SELECT PdDistrict, Date, DayOfWeek, COUNT(*) AS Count \
                         FROM sf_crime WHERE DayOfWeek = 'Sunday' AND -122.4313 <= X AND X <= -122.4213 AND 37.7540 <= Y AND Y<= 37.7740 \
                         GROUP BY PdDistrict, Date, DayOfWeek ORDER BY Date")

display(crimeSunSFDT)

PdDistrict Date DayOfWeek Count NORTHERN 01/01/2006 Sunday 4 SOUTHERN 01/01/2006 Sunday 3 MISSION 01/01/2006 Sunday 16 NORTHERN 01/01/2012 Sunday 1 MISSION 01/01/2012 Sunday 13 MISSION 01/01/2017 Sunday 24 NORTHERN 01/01/2017 Sunday 6 SOUTHERN 01/01/2017 Sunday 1 NORTHERN 01/02/2005 Sunday 4 MISSION 01/02/2005 Sunday 4 MISSION 01/02/2011 Sunday 2 NORTHERN 01/02/2011 Sunday 2 SOUTHERN 01/02/2011 Sunday 3 SOUTHERN 01/03/2010 Sunday 1 MISSION 01/03/2010 Sunday 11 NORTHERN 01/03/2010 Sunday 5 NORTHERN 01/03/2016 Sunday 3 MISSION 01/03/2016 Sunday 6 MISSION 01/04/2004 Sunday 6 SOUTHERN 01/04/2004 Sunday 1 NORTHERN 01/04/2004 Sunday 7 MISSION 01/04/2009 Sunday 8 NORTHERN 01/04/2009 Sunday 2 MISSION 01/04/2015 Sunday 12 MISSION 01/05/2003 Sunday 9 NORTHERN 01/05/2003 Sunday 2 SOUTHERN 01/05/2003 Sunday 3 MISSION 01/05/2014 Sunday 17 SOUTHERN 01/05/2014 Sunday 3 NORTHERN 01/05/2014 Sunday 1 MISSION 01/06/2008 Sunday 6 NORTHERN 01/06/2008 Sunday 1 MISSION 01/06/2013 Sunday 7 SOUTHERN 01/06/2013 Sunday 1 NORTHERN 01/06/2013 Sunday 1 NORTHERN 01/07/2007 Sunday 5 MISSION 01/07/2007 Sunday 14 SOUTHERN 01/07/2007 Sunday 1 MISSION 01/07/2018 Sunday 12 SOUTHERN 01/07/2018 Sunday 1 NORTHERN 01/07/2018 Sunday 2 SOUTHERN 01/08/2006 Sunday 1 NORTHERN 01/08/2006 Sunday 4 MISSION 01/08/2006 Sunday 11 MISSION 01/08/2012 Sunday 13 NORTHERN 01/08/2012 Sunday 1 SOUTHERN 01/08/2012 Sunday 1 MISSION 01/08/2017 Sunday 10 SOUTHERN 01/09/2005 Sunday 2 NORTHERN 01/09/2005 Sunday 4 MISSION 01/09/2005 Sunday 10 MISSION 01/09/2011 Sunday 11 NORTHERN 01/09/2011 Sunday 2 SOUTHERN 01/09/2011 Sunday 6 NORTHERN 01/10/2010 Sunday 2 MISSION 01/10/2010 Sunday 6 SOUTHERN 01/10/2010 Sunday 4 SOUTHERN 01/10/2016 Sunday 2 MISSION 01/10/2016 Sunday 9 MISSION 01/11/2004 Sunday 9 SOUTHERN 01/11/2004 Sunday 2 NORTHERN 01/11/2004 Sunday 5 NORTHERN 01/11/2009 Sunday 3 SOUTHERN 01/11/2009 Sunday 3 MISSION 01/11/2009 Sunday 11 NORTHERN 01/11/2015 Sunday 2 MISSION 01/11/2015 Sunday 7 SOUTHERN 01/12/2003 Sunday 4 MISSION 01/12/2003 Sunday 11 NORTHERN 01/12/2003 Sunday 5 MISSION 01/12/2014 Sunday 10 SOUTHERN 01/12/2014 Sunday 2 NORTHERN 01/12/2014 Sunday 4 MISSION 01/13/2008 Sunday 5 NORTHERN 01/13/2008 Sunday 4 SOUTHERN 01/13/2008 Sunday 1 NORTHERN 01/13/2013 Sunday 2 MISSION 01/13/2013 Sunday 10 NORTHERN 01/14/2007 Sunday 2 MISSION 01/14/2007 Sunday 4 NORTHERN 01/14/2018 Sunday 3 MISSION 01/14/2018 Sunday 7 NORTHERN 01/15/2006 Sunday 1 MISSION 01/15/2006 Sunday 3 NORTHERN 01/15/2012 Sunday 1 MISSION 01/15/2012 Sunday 4 MISSION 01/15/2017 Sunday 26 NORTHERN 01/15/2017 Sunday 3 SOUTHERN 01/15/2017 Sunday 2 NORTHERN 01/16/2005 Sunday 1 MISSION 01/16/2005 Sunday 5 SOUTHERN 01/16/2005 Sunday 2 NORTHERN 01/16/2011 Sunday 2 MISSION 01/16/2011 Sunday 8 SOUTHERN 01/16/2011 Sunday 2 SOUTHERN 01/17/2010 Sunday 3 NORTHERN 01/17/2010 Sunday 7 MISSION 01/17/2010 Sunday 9 MISSION 01/17/2016 Sunday 16 SOUTHERN 01/17/2016 Sunday 3 NORTHERN 01/17/2016 Sunday 6 MISSION 01/18/2004 Sunday 7 MISSION 01/18/2009 Sunday 14 NORTHERN 01/18/2009 Sunday 2 MISSION 01/18/2015 Sunday 10 MISSION 01/19/2003 Sunday 10 SOUTHERN 01/19/2003 Sunday 1 NORTHERN 01/19/2003 Sunday 6 MISSION 01/19/2014 Sunday 11 SOUTHERN 01/19/2014 Sunday 1 NORTHERN 01/19/2014 Sunday 5 MISSION 01/20/2008 Sunday 13 NORTHERN 01/20/2008 Sunday 7 NORTHERN 01/20/2013 Sunday 2 MISSION 01/20/2013 Sunday 10 NORTHERN 01/21/2007 Sunday 3 SOUTHERN 01/21/2007 Sunday 1 MISSION 01/21/2007 Sunday 4 NORTHERN 01/21/2018 Sunday 1 MISSION 01/21/2018 Sunday 18 MISSION 01/22/2006 Sunday 9 NORTHERN 01/22/2006 Sunday 2 MISSION 01/22/2012 Sunday 7 SOUTHERN 01/22/2012 Sunday 2 NORTHERN 01/22/2012 Sunday 3 SOUTHERN 01/22/2017 Sunday 3 MISSION 01/22/2017 Sunday 1 NORTHERN 01/22/2017 Sunday 5 NORTHERN 01/23/2005 Sunday 1 MISSION 01/23/2005 Sunday 9 NORTHERN 01/23/2011 Sunday 4 MISSION 01/23/2011 Sunday 14 NORTHERN 01/24/2010 Sunday 1 MISSION 01/24/2010 Sunday 14 MISSION 01/24/2016 Sunday 6 SOUTHERN 01/24/2016 Sunday 2 NORTHERN 01/25/2004 Sunday 7 MISSION 01/25/2004 Sun

In [14]:
# Analysis the number of crime in each month of 2015, 2016, 2017, 2018. Then, give my insights for the output results. What is the business impact for my result?

crimeTotal = spark.sql("SELECT SUBSTRING(Date,1,2) AS Month, SUBSTRING(Date,7) AS Year, COUNT(*) AS Count \
                       FROM sf_crime GROUP BY Year, Month HAVING Year in (2015, 2016,2017, 2018) ORDER BY Year, Month")

display(crimeTotal)


Month Year Count 01 2015 13606 02 2015 12329 03 2015 13929 04 2015 12959 05 2015 13729 06 2015 13304 07 2015 13365 08 2015 13730 09 2015 12896 10 2015 13147 11 2015 12091 12 2015 11441 01 2016 12967 02 2016 12106 03 2016 12380 04 2016 12328 05 2016 12732 06 2016 12094 07 2016 12191 08 2016 12471 09 2016 12499 10 2016 13388 11 2016 12720 12 2016 13007 01 2017 13084 02 2017 12192 03 2017 13711 04 2017 12941 05 2017 13267 06 2017 12605 07 2017 13171 08 2017 12872 09 2017 12684 10 2017 13355 11 2017 12347 12 2017 12544 01 2018 12031 02 2018 9947 03 2018 10740 04 2018 10306 05 2018 3644

For the output of result, we can see that the number of crimes in 2015, 2016, 2017 are higher than 2018 constantly. They alomost above 12K crimes in each month.

Potential Business Impact: Some shops like jewelry shop would have a safer environment from 2018 since there is less theft crime reported in San Francisco.

In [16]:
# Analysis the number of crime w.r.t the hour in certain day like 2015/12/15, 2016/12/15, 2017/12/15. Give my travel suggestion to visit SF.

hourCrime = spark.sql("SELECT SUBSTRING(Time,1,2) AS Hour, SUBSTRING(Date,1,5) AS Day, SUBSTRING(Date,7) AS Year, COUNT(*) AS Count \
                       FROM sf_crime \
                       WHERE SUBSTRING(Date,1,5) = '12/15' GROUP BY Hour, Day, Year Having Year in (2015,2016,2017) ORDER BY Hour")

display(hourCrime)


Hour Day Year Count 00 12/15 2017 30 00 12/15 2015 15 00 12/15 2016 22 01 12/15 2016 10 01 12/15 2015 6 01 12/15 2017 12 02 12/15 2016 12 02 12/15 2017 8 02 12/15 2015 5 03 12/15 2016 9 03 12/15 2017 4 03 12/15 2015 4 04 12/15 2017 9 04 12/15 2015 10 04 12/15 2016 1 05 12/15 2015 3 05 12/15 2016 6 05 12/15 2017 2 06 12/15 2016 5 06 12/15 2015 4 06 12/15 2017 6 07 12/15 2016 3 07 12/15 2015 8 07 12/15 2017 11 08 12/15 2016 24 08 12/15 2017 26 08 12/15 2015 12 09 12/15 2015 10 09 12/15 2016 23 09 12/15 2017 14 10 12/15 2015 26 10 12/15 2017 27 10 12/15 2016 13 11 12/15 2017 26 11 12/15 2015 15 11 12/15 2016 17 12 12/15 2015 35 12 12/15 2016 29 12 12/15 2017 24 13 12/15 2017 10 13 12/15 2015 13 13 12/15 2016 18 14 12/15 2016 19 14 12/15 2015 27 14 12/15 2017 12 15 12/15 2016 13 15 12/15 2015 14 15 12/15 2017 27 16 12/15 2017 31 16 12/15 2016 17 16 12/15 2015 31 17 12/15 2017 28 17 12/15 2015 18 17 12/15 2016 20 18 12/15 2016 38 18 12/15 2017 29 18 12/15 2015 18 19 12/15 2017 31 19 12/15 2016 31 19 12/15 2015 26 20 12/15 2016 27 20 12/15 2015 13 20 12/15 2017 19 21 12/15 2017 20 21 12/15 2016 16 21 12/15 2015 13 22 12/15 2017 37 22 12/15 2016 22 22 12/15 2015 13 23 12/15 2017 28 23 12/15 2016 16 23 12/15 2015 10

By looking the output result, the number of crimes is pretty in the midnight around averge 22 crimes. The lowest crime number is around from 3AM to 5AM. During the noon and afternoon, there is an gradual increase in crime during December. In addition, during 6PM, the number of crimes increse sharply. Thus, the crime rate is constanly high during the evenings until late evening.

In [18]:
# Step 1: Find out the top-3 danger district
# Step 2: Find out the crime event w.r.t category and time(hour) from the result of step 1
# Give advice to distribute the police based on my analysis results

# Top-3 danger district
crimeDistrictRank = spark.sql("SELECT PdDistrict, COUNT(*) AS CrimeNumber \
                               FROM sf_crime GROUP BY PdDistrict ORDER BY CrimeNumber DESC \
                               LIMIT 3")

display(crimeDistrictRank)

PdDistrict CrimeNumber SOUTHERN 399785 MISSION 300076 NORTHERN 272713

In [19]:
# The crime event w.r.t category and time(hour)
crimeDistrictHour = spark.sql("SELECT PdDistrict, SUBSTRING(Time,1,2) AS Hour, COUNT(*) AS CrimeNumber \
                       FROM sf_crime \
                       WHERE PdDistrict IN ('SOUTHERN', 'MISSION', 'NORTHERN') \
                       GROUP BY PdDistrict, Hour ORDER BY PdDistrict, Hour, CrimeNumber")

display(crimeDistrictHour)


PdDistrict Hour CrimeNumber MISSION 00 16797 MISSION 01 11125 MISSION 02 8930 MISSION 03 5599 MISSION 04 3920 MISSION 05 3089 MISSION 06 4791 MISSION 07 7449 MISSION 08 10647 MISSION 09 11453 MISSION 10 12178 MISSION 11 12731 MISSION 12 16601 MISSION 13 13346 MISSION 14 13728 MISSION 15 14879 MISSION 16 16296 MISSION 17 17486 MISSION 18 18425 MISSION 19 16883 MISSION 20 15743 MISSION 21 15342 MISSION 22 16442 MISSION 23 16196 NORTHERN 00 14472 NORTHERN 01 9291 NORTHERN 02 7809 NORTHERN 03 5450 NORTHERN 04 3729 NORTHERN 05 3287 NORTHERN 06 4220 NORTHERN 07 6263 NORTHERN 08 9535 NORTHERN 09 10194 NORTHERN 10 11022 NORTHERN 11 11004 NORTHERN 12 15457 NORTHERN 13 12212 NORTHERN 14 12690 NORTHERN 15 13706 NORTHERN 16 14319 NORTHERN 17 15816 NORTHERN 18 17508 NORTHERN 19 16503 NORTHERN 20 14819 NORTHERN 21 14406 NORTHERN 22 15008 NORTHERN 23 13993 SOUTHERN 00 20171 SOUTHERN 01 11203 SOUTHERN 02 9295 SOUTHERN 03 5712 SOUTHERN 04 4009 SOUTHERN 05 3447 SOUTHERN 06 5646 SOUTHERN 07 9379 SOUTHERN 08 14341 SOUTHERN 09 16608 SOUTHERN 10 17738 SOUTHERN 11 18196 SOUTHERN 12 23536 SOUTHERN 13 21067 SOUTHERN 14 21510 SOUTHERN 15 22380 SOUTHERN 16 23438 SOUTHERN 17 25276 SOUTHERN 18 26112 SOUTHERN 19 23806 SOUTHERN 20 20610 SOUTHERN 21 18809 SOUTHERN 22 19202 SOUTHERN 23 18294

1. The top three most dangerours districts are SOUTHERN(399785), MISSION(300076), NORTHERN(272713). 
2. By looking through the houly crimes, SOUTHERN has more over crimes than the other two distict, which needs to have more police power covered. 
3. The number of crimes are lowest around 5AM, and during the noon, there is an increase in crime numbers. So the police force shoud be out during noon and before 8pm.

In [21]:
# For different category of crime, find the percentage of resolution. Based on the output, give my hints to adjust the policy.
res = spark.sql("SELECT Category, Resolution, Count(*) AS Count \
                       FROM sf_crime GROUP BY Category, Resolution ORDER BY Category, Resolution")

res.createOrReplaceTempView("resolution")

In [22]:
percentOfRes = spark.sql("SELECT Category, Solved, Unsolved, Solved/(Solved + Unsolved) AS PercentageOfSolved \
                      FROM (SELECT Category, SUM (CASE WHEN Resolution in ('NONE') THEN Count ELSE 0 END) AS Unsolved, \
                      SUM (CASE WHEN Resolution in ('UNFOUNDED') THEN Count ELSE 0 END) as Solved \
                      FROM resolution GROUP BY Category) ORDER BY PercentageOfSolved DESC")

display(percentOfRes)

Category Solved Unsolved PercentageOfSolved SEX OFFENSES, FORCIBLE 504 6845 0.06858075928697782 FAMILY OFFENSES 43 680 0.059474412171507604 SEX OFFENSES, NON FORCIBLE 13 212 0.057777777777777775 VEHICLE THEFT 4789 115980 0.03965421589977561 NON-CRIMINAL 7404 184858 0.03850994996411147 SUSPICIOUS OCC 2573 70986 0.03497872456123656 MISSING PERSON 1014 30289 0.03239306136792001 PORNOGRAPHY/OBSCENE MAT 1 33 0.029411764705882353 PROSTITUTION 25 850 0.02857142857142857 FORGERY/COUNTERFEITING 393 14449 0.02647891119795176 WEAPON LAWS 156 6070 0.025056215868936717 WARRANTS 132 5482 0.023512646954043464 KIDNAPPING 56 2604 0.021052631578947368 EMBEZZLEMENT 46 2147 0.020975832193342453 DRUG/NARCOTIC 219 10271 0.020877025738798856 OTHER OFFENSES 1730 87844 0.01931364011878447 DRIVING UNDER THE INFLUENCE 6 317 0.018575851393188854 GAMBLING 2 114 0.017241379310344827 DISORDERLY CONDUCT 60 3458 0.017055144968732235 RUNAWAY 32 1845 0.017048481619605753 SECONDARY CODES 249 14675 0.016684534977217905 DRUNKENNESS 28 1725 0.015972618368511125 BAD CHECKS 12 744 0.015873015873015872 BRIBERY 5 338 0.014577259475218658 EXTORTION 8 577 0.013675213675213675 STOLEN PROPERTY 17 1439 0.011675824175824176 ASSAULT 1275 113747 0.011084835944428023 TRESPASS 68 6287 0.01070023603461841 LOITERING 3 299 0.009933774834437087 ROBBERY 408 43082 0.009381467003908945 SUICIDE 9 982 0.009081735620585268 FRAUD 285 32057 0.008812070991280688 ARSON 25 3174 0.007814942169427946 BURGLARY 545 76653 0.007059768387782067 LIQUOR LAWS 3 451 0.006607929515418502 VANDALISM 437 101890 0.004270622611822881 LARCENY/THEFT 1221 437927 0.0027803838341515846 RECOVERED VEHICLE 3 8113 3.696402168555939E-4 TREA 0 6 0.0

As the result shown above, we can see that the top three crime resolution issue involves sex offenses, vehicle theft and family offenses under the influence. However, the lowest resolution rate goes to larcency/theft, recovered vechicle and vandalism cases. Thus, we need to adjust policy and give more attention to property theft and vehicle related issue.

- For this exploratory data project, my main goal is to look at the crime data at San Fransisco from 2003 to 2008 and gain some insights into trends of crime and in order to provide some police power, policy adjustment and tourism advices.

- The dataset on SF crime records over two million data points and with some features including IncidntNum, Category, Descript, DayOfWeek and so on. With such a big dataset, I performed it based on Spark SQL and Dataframe.

- When I used Spark SQL and Spark Dataframe API to conduct my project, I learn about the most dangerous crime districts, trend of yearly basis, monthly basis and hourly basis. I also looked at the resolution rate of different categories of crimes to see how to handle it well over the time.